src/qml/nodemodel.cpp

```diff
@@ -28,12 +29,87 @@ void NodeModel::setBlockTipHeight(int new_height)
 
 void NodeModel::setVerificationProgress(double new_progress)
 {
+    QDateTime currentDate = QDateTime::currentDateTime();
+
+    // keep a vector of samples of verification progress at height
+    m_block_process_time.push_front(qMakePair(int(currentDate.toMSecsSinceEpoch()), new_progress));
+
+    // show progress speed if we have more than one sample
+    if (m_block_process_time.size() >= 2) {
+        double progressDelta = 0; //
+        int timeDelta = 0; //
+        int remainingMSecs = 0;
+        double remainingProgress = 1.0 - new_progress; //
+        for (int i = 1; i < m_block_process_time.size(); i++) {
+            QPair<int, double> sample = m_block_process_time[i];
+
+            // take first sample after 500 seconds or last available one
+            if (sample.first < (currentDate.toMSecsSinceEpoch() - 500 * 1000) || i == m_block_process_time.size() - 1) {
+                progressDelta = m_block_process_time[0].second - sample.second;
+                timeDelta = m_block_process_time[0].first - sample.first;
+                remainingMSecs = (progressDelta > 0) ? remainingProgress / progressDelta * timeDelta : -1;
+                break;
+            }
+        }
+        if(remainingMSecs > 0 && m_block_process_time.count() % 1000 == 0) {
+            m_remaining_time = remainingMSecs;
+
+            Q_EMIT remainingTimeChanged();
+        }
+        static const int MAX_SAMPLES = 5000;
+        if (m_block_process_time.count() > MAX_SAMPLES) {
+            m_block_process_time.remove(1, m_block_process_time.count() - 1);
+        }
+    }
+
     if (new_progress != m_verification_progress) {
         m_verification_progress = new_progress;
         Q_EMIT verificationProgressChanged();
     }
 }
 
+bool NodeModel::setBlockTimeList(int new_block_time)
+{
+    int currentTime = QDateTime::currentDateTime().toSecsSinceEpoch();
+    int sec_in_12_hours = 12 * 60 * 60;
+    int time_at_12th_hour = currentTime - currentTime % sec_in_12_hours;
+
+    if (new_block_time < time_at_12th_hour) {
+        return false;
+    }
+
+    double ratio_of_12_hour_passed = (new_block_time - time_at_12th_hour) / (double)sec_in_12_hours;
+
+    m_block_time_list.push_back(ratio_of_12_hour_passed);
+    Q_EMIT blockTimeListChanged();
+    return true;
+}
+
+void NodeModel::setBlockTimeListInitial(const CBlockIndex* pblockindex)
```

Comment:
Yeah this is definitely too low level for us to be passing around here

Reply:
Let me experiment with `interfaces::chain` and devise an approach that avoids low-level object handling.